// ngx-ripple-effect/esm2015/lib/ngx-ripple.module.js
import { NgModule } from '@angular/core';
import { NgxRippleComponent } from './ngx-ripple.component';
import * as i0 from "@angular/core";
export class NgxRippleModule {
}
NgxRippleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.7", ngImport: i0, type: NgxRippleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxRippleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.7", ngImport: i0, type: NgxRippleModule, declarations: [NgxRippleComponent], exports: [NgxRippleComponent] });
NgxRippleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.7", ngImport: i0, type: NgxRippleModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.7", ngImport: i0, type: NgxRippleModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [NgxRippleComponent],
                    exports: [NgxRippleComponent]
                }]
        }] });
