// @ngld/canvas/esm2015/lib/long-press.directive.js
import { Directive, Output, EventEmitter, HostBinding, HostListener } from '@angular/core';
export class LongPressDirective {
    constructor() {
        this.longPress = new EventEmitter();
        this.longPressing = new EventEmitter();
    }
    get clsPress() { return this.isPressing; }
    get clsLongPress() { return this.isLongPressing; }
    onMouseDown(event) {
        this.isPressing = true;
        this.isLongPressing = false;
        this.timeout = setTimeout(() => {
            this.isLongPressing = true;
            this.longPress.emit(event);
            this.interval = setInterval(() => {
                this.longPressing.emit(event);
            }, 50);
        }, 500);
    }
    endPress() {
        clearTimeout(this.timeout);
        clearInterval(this.interval);
        this.isLongPressing = false;
        this.isPressing = false;
    }
}
LongPressDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngldLongPress]'
            },] }
];
LongPressDirective.propDecorators = {
    longPress: [{ type: Output }],
    longPressing: [{ type: Output }],
    clsPress: [{ type: HostBinding, args: ['class.press',] }],
    clsLongPress: [{ type: HostBinding, args: ['class.longpress',] }],
    onMouseDown: [{ type: HostListener, args: ['touchstart', ['$event'],] }, { type: HostListener, args: ['mousedown', ['$event'],] }],
    endPress: [{ type: HostListener, args: ['touchend',] }, { type: HostListener, args: ['mouseup',] }, { type: HostListener, args: ['mouseleave',] }]
};
