// tree-ngx/esm2020/tree-ngx.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TreeNgxComponent } from './src/tree-ngx/tree-ngx.component';
import { NodeComponent } from './src//node/node.component';
import { TreeService } from './src//service/tree-service';
import { NodeNameComponent } from './src/node-name/node-name.component';
import { NodeIconWrapperComponent } from './src/node-icon-wrapper/node-icon-wrapper.component';
import * as i0 from "@angular/core";
export class TreeNgxModule {
}
TreeNgxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TreeNgxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TreeNgxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TreeNgxModule, declarations: [TreeNgxComponent,
        NodeComponent,
        NodeNameComponent,
        NodeIconWrapperComponent], imports: [CommonModule], exports: [TreeNgxComponent] });
TreeNgxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TreeNgxModule, providers: [
        TreeService
    ], imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: TreeNgxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        TreeNgxComponent,
                        NodeComponent,
                        NodeNameComponent,
                        NodeIconWrapperComponent,
                    ],
                    providers: [
                        TreeService
                    ],
                    exports: [
                        TreeNgxComponent
                    ]
                }]
        }] });
