// @ngld/canvas/esm2015/lib/info-panel-component/info-panel.component.js
import { Component } from '@angular/core';
import { ActiveIndividualStoreService } from '../active-individual-store.service';
import { ColorProviderService } from '../color-provider.service';
export class InfoPanelComponent {
    constructor(activeIndividualStore, colorGetter) {
        this.activeIndividualStore = activeIndividualStore;
        this.colorGetter = colorGetter;
    }
}
InfoPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngld-info-panel',
                template: "<div *ngIf=\"activeIndividualStore.activeIndividual$ | async as active;\">\n  <ng-container *ngTemplateOutlet=\"infoPanel; context: {data: active}\"></ng-container>\n</div>\n\n\n<ng-template #infoPanel let-data='data'>\n  <a href=\"#\" class=\"class\" [style.backgroundColor]=\"colorGetter.nextColorByName(data.labels[0])\">\n    <strong> {{ data.labels || data.type }} </strong>\n  </a>\n  <a href=\"#\" class=\"property\">\n    <strong> {{ data.id }} </strong>\n  </a>\n</ng-template>\n",
                styles: [":host{font-size:16px;padding:10px}a{border:1px solid;display:inline-block;font-size:14px;line-height:1.428571429;margin-left:5px;margin-top:5px;padding:6px 12px}a.class{color:#fff}a.property{background-color:#fff;border-color:#ccc;color:#333}a.btn{margin-left:5px;margin-top:5px;opacity:1}a.info{background-color:#a5abb6;border:1px solid #9aa1ac;color:#fff}\n"]
            },] }
];
InfoPanelComponent.ctorParameters = () => [
    { type: ActiveIndividualStoreService },
    { type: ColorProviderService }
];
