// tree-ngx/esm2020/src/node/node.component.mjs
import { Component, ViewChild, Input } from '@angular/core';
import { NodeSelectedState } from '../model/node-selected-state';
import * as i0 from "@angular/core";
import * as i1 from "../service/tree-service";
import * as i2 from "../node-icon-wrapper/node-icon-wrapper.component";
import * as i3 from "../node-name/node-name.component";
import * as i4 from "@angular/common";
export class NodeComponent {
    constructor(treeService) {
        this.treeService = treeService;
        this._this = this;
    }
    ngOnChanges(changes) {
        if (changes.selectedState) {
            this.selectedStateChanged();
        }
    }
    ngAfterViewInit() {
        this.selectedStateChanged();
    }
    selectedStateChanged() {
        if (this.nodeCheckbox) {
            if (this.selectedState === NodeSelectedState.indeterminate) {
                this.nodeCheckbox.nativeElement.indeterminate = true;
            }
            else {
                this.nodeCheckbox.nativeElement.indeterminate = false;
            }
            if (this.selectedState === NodeSelectedState.checked) {
                this.nodeCheckbox.nativeElement.checked = true;
            }
            else if (this.selectedState === NodeSelectedState.unChecked) {
                this.nodeCheckbox.nativeElement.checked = false;
            }
        }
    }
    checkBoxClick() {
        this.treeService.checkBoxClick(this.state);
        setTimeout(() => {
            this.treeService.reEvaluateSelectedState(this.state);
        }, 1);
    }
}
NodeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: NodeComponent, deps: [{ token: i1.TreeService }], target: i0.ɵɵFactoryTarget.Component });
NodeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.2", type: NodeComponent, selector: "node", inputs: { state: "state", selectedState: "selectedState", nodeNameTemplate: "nodeNameTemplate", nodeCollapsibleTemplate: "nodeCollapsibleTemplate" }, viewQueries: [{ propertyName: "nodeCheckbox", first: true, predicate: ["nodeCheckbox"], descendants: true }], usesOnChanges: true, ngImport: i0, template: "<div class=\"node\">\r\n  <div class=\"node-container\">\r\n    <node-icon-wrapper [state]=\"state\" [nodeCollapsibleTemplate]=\"nodeCollapsibleTemplate\"></node-icon-wrapper>\r\n\r\n    <input #nodeCheckbox class=\"node-checkbox\" *ngIf=\"state.showCheckBox\" (click)=\"checkBoxClick()\" [disabled]=\"state.disabled\" [checked]=\"state.selected\" type=\"checkbox\" />\r\n    <node-name [state]=\"state\" [nodeNameTemplate]=\"nodeNameTemplate\"></node-name>\r\n  </div>\r\n  <div *ngIf=\"state.nodeItem.children\" class=\"collapsible-wrapper\" [class.collapsed]=\"!state.expanded && state.nodeItem.children.length > 0 && state.filteredChildren.length > 0\">\r\n    <div class=\"collapsible\">\r\n      <node class=\"node-offset\" [state]=\"child\" [selectedState]=\"child.selectedState\" [nodeCollapsibleTemplate]=\"nodeCollapsibleTemplate\" [nodeNameTemplate]=\"nodeNameTemplate\" *ngFor=\"let child of state.filteredChildren\"></node>\r\n    </div>\r\n  </div>\r\n</div>\r\n", components: [{ type: i2.NodeIconWrapperComponent, selector: "node-icon-wrapper", inputs: ["state", "nodeCollapsibleTemplate"] }, { type: i3.NodeNameComponent, selector: "node-name", inputs: ["state", "nodeNameTemplate"] }, { type: NodeComponent, selector: "node", inputs: ["state", "selectedState", "nodeNameTemplate", "nodeCollapsibleTemplate"] }], directives: [{ type: i4.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: NodeComponent, decorators: [{
            type: Component,
            args: [{ selector: 'node', template: "<div class=\"node\">\r\n  <div class=\"node-container\">\r\n    <node-icon-wrapper [state]=\"state\" [nodeCollapsibleTemplate]=\"nodeCollapsibleTemplate\"></node-icon-wrapper>\r\n\r\n    <input #nodeCheckbox class=\"node-checkbox\" *ngIf=\"state.showCheckBox\" (click)=\"checkBoxClick()\" [disabled]=\"state.disabled\" [checked]=\"state.selected\" type=\"checkbox\" />\r\n    <node-name [state]=\"state\" [nodeNameTemplate]=\"nodeNameTemplate\"></node-name>\r\n  </div>\r\n  <div *ngIf=\"state.nodeItem.children\" class=\"collapsible-wrapper\" [class.collapsed]=\"!state.expanded && state.nodeItem.children.length > 0 && state.filteredChildren.length > 0\">\r\n    <div class=\"collapsible\">\r\n      <node class=\"node-offset\" [state]=\"child\" [selectedState]=\"child.selectedState\" [nodeCollapsibleTemplate]=\"nodeCollapsibleTemplate\" [nodeNameTemplate]=\"nodeNameTemplate\" *ngFor=\"let child of state.filteredChildren\"></node>\r\n    </div>\r\n  </div>\r\n</div>\r\n" }]
        }], ctorParameters: function () { return [{ type: i1.TreeService }]; }, propDecorators: { nodeCheckbox: [{
                type: ViewChild,
                args: ['nodeCheckbox', { static: false }]
            }], state: [{
                type: Input
            }], selectedState: [{
                type: Input
            }], nodeNameTemplate: [{
                type: Input
            }], nodeCollapsibleTemplate: [{
                type: Input
            }] } });
