// tree-ngx/esm2020/src/util/util.mjs
import { NodeSelectedState } from '../model/node-selected-state';
import { TreeMode } from '../model/tree-mode';
export class TreeUtil {
    static initState(parent, nodeItem, options) {
        const nodeState = {
            parent: parent,
            children: [],
            filteredChildren: [],
            hasFilteredChildren: false,
            nodeItem: nodeItem,
            expanded: nodeItem.expanded === false ? false : true,
            disabled: nodeItem.disabled === true ? true : false,
            markSelected: this.getMarkSelected(nodeItem, options),
            selectedState: NodeSelectedState.unChecked,
            selected: false,
            showCheckBox: options.checkboxes
        };
        return nodeState;
    }
    static getMarkSelected(nodeItem, options) {
        if (((!nodeItem.children || nodeItem.children.length === 0) || options.mode === TreeMode.SingleSelect) && !options.checkboxes) {
            return true;
        }
        else {
            return false;
        }
    }
}
